### ISSUE TITLE
Add optional `search_path` to `postgresql_role` resource

### ISSUE THREAD

**Issue by dhermes** [terraform-providers/terraform-provider-postgresql#90]
I'm happy to take this on but wanted to make sure this was a feature the maintainers would be interested in adding before I start coding. The idea would be that it'd run

```
ALTER ROLE {postgresql_role.FOO.name}
  SET search_path TO {postgresql_role.FOO.search_path};
```

after creating the user. I'm not sure what the general approach to such a feature would be, i.e. 

- Do we need to make sure the schema in `search_path` exists?
- Do we need to make sure the role has access to the schema in `search_path`?
- Do we **always** run `ALTER ROLE ... SET search_path TO ...` and just use `public` as the default value or conditionally run it if explicitly set on the `postgresql_role`?

### Terraform Version

```
$ terraform -v
Terraform v0.12.3

```

### Affected Resource(s)

`postgresql_role`

**Comment by cyrilgdn** [comment]
Hi @dhermes , thank you for opening this issue.

Yes we would be interested by this feature. There's actually 2 ways to define `search_path` for a role:

*  ``ALTER ROLE _role_name_ SET search_path TO  _schema_name_`` which set it at the server level.
*  ``ALTER ROLE _role_name_ IN DATABASE _dbname_ SET search_path TO  _schema_name_`` which set it at a database level.

The second one will be a bit more complex to implement I guess as we need to be able to configure it for multiple databases so itt needs to have a more complex field.
If you need only the first one for now (so define a global `search_path` for a role) you can start to implement this with a simple string field (or maybe a list of string would be better I'm not sure?).

In this case:

> Do we need to make sure the schema in search_path exists?

I would say no as it's configured for all databases and schema may not exist in all databases.

> Do we need to make sure the role has access to the schema in search_path

No for the same reason.

> Do we always run ALTER ROLE ... SET search_path TO ... and just use public as the default value or conditionally run it if explicitly set on the postgresql_role?

I would say no for compatibility reason. User may already have defined some `search_path` on database for all roles (with `ALTER DATABASE xxx SET search_path=yyy`) so we don't want to change this behavior by forcing a `search_path` on each role as role `search_path` takes precedence on the database one.

**Comment by dhermes** [comment]
> If you need only the first one for now (so define a global search_path for a role) you can start to implement this with a simple string field (or maybe a list of string would be better I'm not sure?)

I agree with both of these 😀 

- I'll make it global (for now)
- I'll use a list rather than a string

Thanks for your time!

**Comment by dhermes** [comment]
So @cyrilgdn I wanted to talk about how the `search_path` fits into [`resourcePostgreSQLRoleReadImpl()`][1]. I'd like to set the value via

```go
d.Set(roleSearchPathAttr, getSearchPath(roleConfigs))
```

where `roleConfigs` is a `text[]` value from the `rolconfig` column of `pg_catalog.pg_roles`. This is actually quite painful to parse (i.e. it needs a lexer to do it right). For just a small example of the sorts of values that can be set there:

```sql
postgres=# CREATE ROLE foo;
CREATE ROLE
postgres=# ALTER ROLE foo SET search_path = xyz;
ALTER ROLE
postgres=# SELECT rolconfig FROM pg_catalog.pg_roles WHERE rolname = 'foo';
     rolconfig
-------------------
 {search_path=xyz}
(1 row)

postgres=#
postgres=# ALTER ROLE foo SET search_path = ab, cd;
ALTER ROLE
postgres=# SELECT rolconfig FROM pg_catalog.pg_roles WHERE rolname = 'foo';
       rolconfig
------------------------
 {"search_path=ab, cd"}
(1 row)

postgres=#
postgres=# ALTER ROLE foo SET DateStyle = ISO;
ALTER ROLE
postgres=# SELECT rolconfig FROM pg_catalog.pg_roles WHERE rolname = 'foo';
              rolconfig
--------------------------------------
 {"search_path=ab, cd",DateStyle=iso}
(1 row)

postgres=# ALTER ROLE foo RESET DateStyle;
ALTER ROLE
postgres=# ALTER ROLE foo SET search_path = ab, cd, "e-f";
ALTER ROLE
postgres=# SELECT rolconfig FROM pg_catalog.pg_roles WHERE rolname = 'foo';
            rolconfig
---------------------------------
 {"search_path=ab, cd, \"e-f\""}
(1 row)

postgres=# ALTER ROLE foo SET search_path = "e-f "" g";
ALTER ROLE
postgres=# SELECT rolconfig FROM pg_catalog.pg_roles WHERE rolname = 'foo';
           rolconfig
--------------------------------
 {"search_path=\"e-f \"\" g\""}
(1 row)

```

I spent quite a bit time Googling around and there isn't a much better way to get this information. Using `current_schemas()` is a higher quality value but only applies to the current user and only applies to actually created schemas:

```sql
postgres=# SET search_path = xy, "a-b";
SET
postgres=# SELECT current_schemas(false);
 current_schemas 
-----------------
 {}
(1 row)

postgres=# CREATE SCHEMA xy;
CREATE SCHEMA
postgres=# SELECT current_schemas(false);
 current_schemas 
-----------------
 {xy}
(1 row)

postgres=# CREATE SCHEMA "a-b";
CREATE SCHEMA
postgres=# SELECT current_schemas(false);
 current_schemas 
-----------------
 {xy,a-b}
(1 row)

```

I think it may not be feasible to set a default value. If that's the route I go, I can do what is done in [`resourcePostgreSQLRoleUpdate()`][2] with `revokeRoles()` / `grantRoles()` and just always call

```sql
ALTER ROLE foo SET search_path = ...
```

independent of whether a change is needed.

[1]: https://github.com/terraform-providers/terraform-provider-postgresql/blob/b1f67a3af1edb5355d9f7833efb8340a37aaf9ab/postgresql/resource_postgresql_role.go#L423
[2]: https://github.com/terraform-providers/terraform-provider-postgresql/blob/b1f67a3af1edb5355d9f7833efb8340a37aaf9ab/postgresql/resource_postgresql_role.go#L545-L552

**Comment by cyrilgdn** [comment]
Hi @dhermes ,

Sorry for the late answer (I had a couple of weeks vacations).

It can indeed be hard to parse the exact list of schema with all your examples but actually we don't really need to manage them.
As soon as we manage this field, the provider is responsible of its value and its format.

So you choose which format you want to support and you write it and read it like that.
e.g.:: ``schema1, schema2``

In the ``resourcePostgreSQLRoleReadImpl`` function, you can simply do something like:
```
configs := readConfigs(roleConfig)
d.Set(roleSearchPath, strings.Split(configs["search_path"], ", "))
```

When``roleConfig`` is read directly from the ``rolconfig`` field in Postgres and ``readConfigs`` is a func like : (not tested)
```
func readConfigs(array pq.ByteaArray) map[string]string {
	configs := map[string]string{}
	for _, v := range array {
		value := strings.Split(string(v), "=")
		if len(value) > 1 {
			configs[value[0]] = value[1]
		}
	}
	return configs
}
```

And in the ``resourcePostgreSQLRoleCreate`` / ``resourcePostgreSQLRoleUpdate`` you can call a ``setSearchPath`` function which will execute something like:
```
searchPath := d.Get("roles").(*schema.Set).List()
fmt.Sprintf("ALTER ROLE %s SET search_path=%s", rolename, strings.Join(searchPath, ", "))
```
(You probably need to:
 * transform ``searchPath`` to ``[]string`` 
 * Call ``pq.QuoteIdentifier`` on each schema name to protect against SQL injection
 * Manage the case when searchPath is empty so you need to call ``ALTER ROLE %s RESET search_path`` instead of ``SET search_path=xxx``
)

If ``rolconfig`` contains a search_path with another format, it means that someone updated it manually and Terraform will update it with the right format (like if someone update manually the value of any other field).

**Comment by dhermes** [comment]
So `strings.Split(configs["search_path"], ", ")` will not be sufficient. Due to the escaping, it's actually allowed to put a comma in a schema name for some pathological reason. E.g.

```
postgres=# CREATE ROLE foo;
CREATE ROLE
postgres=# ALTER ROLE foo SET search_path = ab, cd, "ab, cd";
ALTER ROLE
postgres=# SELECT rolconfig FROM pg_catalog.pg_roles WHERE rolname = 'foo';
             rolconfig              
------------------------------------
 {"search_path=ab, cd, \"ab, cd\""}
(1 row)

```

Though reading your message I now realize we could just not bother with parsing `configs["search_path"]` and check that the full string matches whatever we'd produce (i.e. some canonical / normalized string including the quoting you mentioned).

I actually wrote a lexer / parser called [`unquoteIdentifierList()` here][1], I just was trying to avoid that level of complexity. LMK what you think about avoiding parsing and just taking the full `string` / `[]byte` as the value we check equality against.

[1]: https://play.golang.org/p/dsR8sPXPgCS

**Comment by cyrilgdn** [comment]
> it's actually allowed to put a comma in a schema name for some pathological reason

Ouch... That hurts :sob: !

> we could just not bother with parsing

Actually we have to if we set this field as a `TypeSet` because in the `resourcePostgreSQLRoleRead` function we need to fill the list with the value read from the database.
Then Terraform will compare the `Set` we've read from the database with the `Set` it read from the `tf` files.

If it can help we could take the decision to forbid comma in schema name (I mean, is there really people doing that in this world :cry: ?). But I guess you'll still have to take care of quotes if there's other special characters in the name.

You can also take a look at the [`DiffSuppressFunc`](https://github.com/hashicorp/terraform/blob/master/helper/schema/schema.go#L107-L114) attribute. Maybe it could help? (But if I remember correctly, the behavior with `TypeSet` is not so helpful :( )

**Comment by NinoFloris** [comment]
Any luck on progression here?

**Comment by Hjdskes** [comment]
@NinoFloris See my pull request, #93 :)

**Comment by cyrilgdn** [comment]
Closed by #93 and released last Friday in v1.3.0

Thanks @Hjdskes @dhermes and @navi86 for your work / research!

**Comment by dhermes** [comment]
:tada:
